crates/autopilot/src/domain/settlement/solution.rs

```diff
@@ -0,0 +1,363 @@
+                    },
+                },
+            ));
+        }
+
+        Ok(Self { trades, auction_id })
+    }
+}
+
+/// Trade flags are encoded in a 256-bit integer field. For more information on
+/// how flags are encoded see:
+/// <https://github.com/cowprotocol/contracts/blob/v1.0.0/src/contracts/libraries/GPv2Trade.sol#L58-L94>
+#[derive(Debug, PartialEq, Eq)]
+pub struct TradeFlags(pub U256);
+
+impl TradeFlags {
+    fn as_u8(&self) -> u8 {
+        self.0.byte(0)
+    }
+
+    pub fn side(&self) -> order::Side {
+        if self.as_u8() & 0b1 == 0 {
+            order::Side::Sell
+        } else {
+            order::Side::Buy
+        }
+    }
+
+    pub fn partially_fillable(&self) -> bool {
+        self.as_u8() & 0b10 != 0
+    }
+
+    pub fn sell_token_balance(&self) -> boundary::SellTokenSource {
+        if self.as_u8() & 0x08 == 0 {
+            boundary::SellTokenSource::Erc20
+        } else if self.as_u8() & 0x04 == 0 {
+            boundary::SellTokenSource::External
+        } else {
+            boundary::SellTokenSource::Internal
+        }
+    }
+
+    pub fn buy_token_balance(&self) -> boundary::BuyTokenDestination {
+        if self.as_u8() & 0x10 == 0 {
+            boundary::BuyTokenDestination::Erc20
+        } else {
+            boundary::BuyTokenDestination::Internal
+        }
+    }
+
+    pub fn signing_scheme(&self) -> boundary::SigningScheme {
+        match (self.as_u8() >> 5) & 0b11 {
+            0b00 => boundary::SigningScheme::Eip712,
+            0b01 => boundary::SigningScheme::EthSign,
+            0b10 => boundary::SigningScheme::Eip1271,
+            0b11 => boundary::SigningScheme::PreSign,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl From<U256> for TradeFlags {
+    fn from(value: U256) -> Self {
+        Self(value)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("transaction calldata is not a settlement")]
+    InvalidSelector,
+    #[error("no auction id found in calldata")]
+    MissingAuctionId,
+    #[error("auction {0} failed encoding: {1}")]
+    Encoding(auction::Id, error::Encoding),
+}
+
+pub mod error {
+    use super::*;
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Encoding {
+        #[error("unable to decode settlement calldata: {0}")]
+        Decoding(#[from] web3::ethabi::Error),
+        #[error("unable to tokenize calldata into expected format: {0}")]
+        Tokenizing(#[from] ethcontract::tokens::Error),
+        #[error("unable to recover order uid: {0}")]
+        OrderUidRecover(#[from] tokenized::Error),
+    }
+
+    impl Encoding {
+        pub fn with(self, auction: auction::Id) -> Error {
+            Error::Encoding(auction, self)
+        }
+    }
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Score {
+        #[error(transparent)]
+        Trade(#[from] trade::Error),
+        #[error(transparent)]
+        Zero(#[from] competition::ZeroScore),
```

Comment:
Can we document these errors? Not sure what a `settlement::solution::error::Score::Trade` should mean. I also don't think they should be collapsed (I know it's convenient to just wrap an existing error type into another one), but I think the correct score error types should be
- Zero
- MissingPrice
- Math

Where trade::Error get translated accordingly.